runtime/compiler/env/J9SharedCache.cpp

```diff
@@ -814,75 +857,150 @@ TR_J9SharedCache::findChainForClass(J9Class *clazz, const char *key, uint32_t ke
+      }
+   else if (!validateSuperClassesInClassChain(clazz, chainPtr, chainEnd))
+      {
+      LOG(1, "\tClass in hierarchy did not match, returning false\n");
+      }
+   else if (!validateInterfacesInClassChain(clazz, chainPtr, chainEnd))
+      {
+      LOG(1, "\tInterface class did not match, returning false\n");
+      }
+   else if (chainPtr != chainEnd)
+      {
+      LOG(1, "\tfinished classes and interfaces, but not at chain end, returning false\n");
+      }
+   else
+      {
+      validationSucceeded = true;
       }
 
-   int32_t numSuperclasses = TR::Compiler->cls.classDepthOf(fe()->convertClassPtrToClassOffset(clazz));
-   for (int32_t index=0; index < numSuperclasses;index++)
+   return validationSucceeded;
+   }
+
+bool
+TR_J9SharedCache::classMatchesCachedVersion(J9Class *clazz, UDATA *chainData)
+   {
+   J9ROMClass *romClass = TR::Compiler->cls.romClassOf(fe()->convertClassPtrToClassOffset(clazz));
+   J9UTF8 * className = J9ROMCLASS_CLASSNAME(romClass);
+   LOG(1, "classMatchesCachedVersion class %p %.*s\n", clazz, J9UTF8_LENGTH(className), J9UTF8_DATA(className));
+
+   uintptr_t classOffsetInCache;
+
+   /* If the pointer isn't the SCC, then return false immmediately
+    * as the map holds offsets into the SCC of romclasses
+    */
+   if (!isPointerInSharedCache(romClass, &classOffsetInCache))
       {
-      J9ROMClass *romClass = TR::Compiler->cls.romClassOfSuperClass(fe()->convertClassPtrToClassOffset(clazz), index);
-      if (!romclassMatchesCachedVersion(romClass, chainPtr, chainEnd))
-         {
-         LOG(5, { log("\tClass in hierarchy did not match, returning false\n"); });
-         return false;
-         }
+      LOG(1, "\tclass not in shared cache, returning false\n");
+      return false;
       }
 
-   J9ITable *interfaceElement = TR::Compiler->cls.iTableOf(fe()->convertClassPtrToClassOffset(clazz));
-   while (interfaceElement)
+   /* Check if the validation of the class chain was previously
+    * performed; if so, return the result of that validation
+    */
+   if (TR::Options::getAOTCmdLineOptions()->getOption(TR_EnableClassChainValidationCaching))
       {
-      J9ROMClass * romClass = TR::Compiler->cls.iTableRomClass(interfaceElement);
-      if (!romclassMatchesCachedVersion(romClass, chainPtr, chainEnd))
+      auto result = getCachedCCVResult(classOffsetInCache);
+      if (result == CCVResult::success)
+         {
+         LOG(1, "\tcached result: validation succeeded\n");
+         return true;
+         }
+      else if (result == CCVResult::failure)
          {
-         LOG(5, { log("\tInterface class did not match, returning false\n"); });
+         LOG(1, "\tcached result: validation failed\n");
          return false;
          }
-      interfaceElement = TR::Compiler->cls.iTableNext(interfaceElement);
+      else
+         {
+         TR_ASSERT_FATAL(result == CCVResult::notYetValidated, "Unknown result cached %d\n", result);
```

Comment:
I don't understand this assert. When `TR_EnableClassChainValidationCaching` is enabled, we either return true or false or crash. Then the code below that does `cacheCCVResult(classOffsetInCache, CCVResult::failure)` is unreachable.

Reply:
`getCachedCCVResult` can return `CCVResult::notYetValidated` if we've never cached the result of the validation yet.

Reply:
The assert is to ensure that if the result returned is not `success` or `failure`, it better be `notYetValidated`.